File: static/CWE-22/python/patched.py
# Remediation for CWE-22: Path Traversal
# Fix: Resolve the requested path and confirm it stays within the allowed base directory.

import os
from flask import Flask, request, abort

app = Flask(__name__)

BASE_DIR = "/var/www/files"


@app.route("/download")
def download():
    filename = request.args.get("file", "")
    base = os.path.realpath(BASE_DIR)
    path = os.path.realpath(os.path.join(base, filename))
    if os.path.commonpath([base, path]) != base:
        abort(403)
    with open(path) as f:
        return f.read()

File: static/CWE-22/javascript/patched.js
// Remediation for CWE-22: Improper Limitation of a Pathname to a Restricted Directory ('Path Traversal')
// Fix: Resolve the requested path and verify it stays contained within the allowed base directory before reading.

const express = require('express');
const fs = require('fs');
const path = require('path');

const app = express();

const BASE_DIR = path.resolve(__dirname, 'files');

app.get('/download', (req, res) => {
  const file = req.query.file;
  const filePath = path.resolve(BASE_DIR, file);
  if (filePath !== BASE_DIR && !filePath.startsWith(BASE_DIR + path.sep)) {
    return res.status(400).send('invalid path');
  }
  fs.readFile(filePath, 'utf8', (err, data) => {
    if (err) return res.status(404).send('not found');
    res.send(data);
  });
});

app.listen(3000);
